pkg/api/server/v1alpha2/auth/impersonation/impersonation.go

```diff
@@ -0,0 +1,217 @@
+				Name:     users[0],
+				Resource: "users",
+				Verb:     "impersonate",
+			})
+		}
+	}
+
+	groups := md.Get(authenticationv1.ImpersonateGroupHeader)
+	hasGroups := len(groups) > 0
+	if hasGroups {
+		for _, group := range groups {
+			i.userInfo.Groups = append(i.userInfo.Groups, group)
+			i.resourceAttributes = append(i.resourceAttributes, authorizationv1.ResourceAttributes{
+				Name:     group,
+				Resource: "groups",
+				Verb:     "impersonate",
+			})
+		}
+	}
+
+	UIDs := md.Get(authenticationv1.ImpersonateUIDHeader)
+	hasUID := len(UIDs) > 0
+	if hasUID {
+		i.userInfo.UID = UIDs[0]
+		i.resourceAttributes = append(i.resourceAttributes, authorizationv1.ResourceAttributes{
+			Group:    authenticationv1.SchemeGroupVersion.Group,
+			Name:     UIDs[0],
+			Resource: "uids",
+			Verb:     "impersonate",
+		})
+	}
+
+	hasExtra := false
+	for name, values := range md {
+		if !strings.HasPrefix(name, strings.ToLower(authenticationv1.ImpersonateUserExtraHeaderPrefix)) {
+			continue
+		}
+
+		hasExtra = true
+		extraKey := unescapeExtraKey(strings.ToLower(name[len(authenticationv1.ImpersonateUserExtraHeaderPrefix):]))
+		i.userInfo.Extra = make(map[string][]string)
+		// Each extra value is a separate resource to check.
+		for _, value := range values {
+			i.userInfo.Extra[extraKey] = append(i.userInfo.Extra[extraKey], value)
+			i.resourceAttributes = append(i.resourceAttributes, authorizationv1.ResourceAttributes{
+				Group:       authenticationv1.SchemeGroupVersion.Group,
+				Name:        value,
+				Resource:    "userextras",
+				Subresource: extraKey,
+				Verb:        "impersonate",
+			})
+		}
+	}
+
+	// When impersonating a non-anonymous user, include the 'system:authenticated' group in the impersonated user info:
+	// - if no groups were specified
+	// - if a group has been specified other than 'system:authenticated'
+	// If 'system:unauthenticated' group has been specified we should not include the 'system:authenticated' group.
+	if hasUser {
+		if i.userInfo.Name != user.Anonymous {
+			addAuthenticated := true
+			for _, group := range i.userInfo.Groups {
+				if group == user.AllAuthenticated || group == user.AllUnauthenticated {
+					addAuthenticated = false
+					break
+				}
+			}
+
+			if addAuthenticated {
+				i.userInfo.Groups = append(i.userInfo.Groups, user.AllAuthenticated)
+			}
+		} else {
+			addUnauthenticated := true
+			for _, group := range i.userInfo.Groups {
+				if group == user.AllUnauthenticated {
+					addUnauthenticated = false
+					break
+				}
+			}
+
+			if addUnauthenticated {
+				i.userInfo.Groups = append(i.userInfo.Groups, user.AllUnauthenticated)
+			}
+		}
+	} else {
+		i.userInfo = nil
+
+		// Impersonate-User header is mandatory.
+		// https://kubernetes.io/docs/reference/access-authn-authz/authentication/#user-impersonation
+		if hasGroups || hasExtra || hasUID {
+			return ErrorImpersonateUserRequired
+		} else {
+			return ErrorNoImpersonationData
+		}
```

Comment:
Should we return an error here? This seems to imply that _all_ requests require impersonation, which is certainly not the case for requests that come in from the watcher.

Reply:
I returned error here so that the `NewImpersonation` function doesn't return an empty or nil `Impersonation` struct without returning any error.
This error will not stop the processing of the requests, because it's handled here.

https://github.com/tektoncd/results/blob/06adf689eb14324be872c43787d6d68861df5af9/pkg/api/server/v1alpha2/auth/rbac.go#L71-L74

The idea here is if no impersonation header is present we, continue processing with a nil `Impersonation` struct, which is checked later while performing impersonation check. But if any other impersonation header is present without having `Impersonate-User` header, then we stop processing of that request. This is the same behaviour in the Kubernetes impersonation filter implementation.